### PR TITLE
ARROW-554: [C++] Add functions to unify dictionary types and arrays

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

File: cpp/src/arrow/builder-dict.cc
Comment (pitrou) [reply cpp/src/arrow/builder-dict.cc]: Probably... but I'm not sure that would be more readable, as I then would have to reproduce the type selection logic in the template instantiation.

Comment (bkietz) [reply cpp/src/arrow/builder-dict.cc]: I'd say that it's more readable to write an expression delineating supported types (see `CanUnify` below) than to list them explicitly in a macro'd switch. Incidentally, why isn't unification of `DecimalType` supported?

<details>

```c++
struct UnifyDictionaryValues {
  MemoryPool* pool;
  std::shared_ptr<DataType> value_type;
  const std::vector<const DictionaryType*>& types;
  std::shared_ptr<Array>* out_values;
  std::vector<std::vector<int32_t>>* out_transpose_maps;

  template <typename T>
  static constexpr bool CanUnify() {
      return std::is_base_of<BinaryType, T>::value
        || (std::is_base_of<FixedWidthType, T>::value
            && !std::is_same<Decimal128Type, T>::value);
  }

  template <typename T>
  Status Visit(const T&, typename std::enable_if<!CanUnify<T>()>* = nullptr) {
      std::stringstream ss;
      ss << "Unification of " << value_type->ToString()
         << " dictionaries is not implemented";
      return Status::NotImplemented(ss.str());
  }

  template <typename T>
  Status Visit(const T&, typename std::enable_if<CanUnify<T>()>* = nullptr) {
    using ArrayType = typename TypeTraits<T>::ArrayType;
    using DictTraits = typename internal::DictionaryTraits<T>;
    using MemoTableType = typename DictTraits::MemoTableType;
    
    MemoTableType memo_table;
    if (out_transpose_maps != nullptr) {
      out_transpose_maps->clear();
      out_transpose_maps->reserve(types.size());
    }
    // Build up the unified dictionary values and the transpose maps
    for (const auto& type : types) {
      const ArrayType& values = checked_cast<const ArrayType&>(*type->dictionary());
      if (out_transpose_maps != nullptr) {
        std::vector<int32_t> transpose_map;
        transpose_map.reserve(values.length());
        for (int64_t i = 0; i < values.length(); ++i) {
          int32_t dict_index = memo_table.GetOrInsert(values.GetView(i));
          transpose_map.push_back(dict_index);
        }
        out_transpose_maps->push_back(std::move(transpose_map));
      } else {
        for (int64_t i = 0; i < values.length(); ++i) {
          memo_table.GetOrInsert(values.GetView(i));
        }
      }
    }
    // Build unified dictionary array
    std::shared_ptr<ArrayData> data;
    RETURN_NOT_OK(DictTraits::GetDictionaryArrayData(pool, value_type, memo_table,
                                                    0 /* start_offset */, &data));
    *out_values = MakeArray(data);
    return Status::OK();
  }
};

Status DictionaryType::Unify(MemoryPool* pool, const std::vector<const DataType*>& types,
                             std::shared_ptr<DataType>* out_type,
                             std::vector<std::vector<int32_t>>* out_transpose_maps) {
  if (types.size() == 0) {
    return Status::Invalid("need at least one input type");
  }
  std::vector<const DictionaryType*> dict_types;
  dict_types.reserve(types.size());
  for (const auto& type : types) {
    if (type->id() != Type::DICTIONARY) {
      return Status::TypeError("input types must be dictionary types");
    }
    dict_types.push_back(checked_cast<const DictionaryType*>(type));
  }
  // XXX Should we check the ordered flag?
  auto value_type = dict_types[0]->dictionary()->type();
  for (const auto& type : dict_types) {
    auto values = type->dictionary();
    if (!values->type()->Equals(value_type)) {
      return Status::TypeError("input types have different value types");
    }
    if (values->null_count() != 0) {
      return Status::TypeError("input types have null values");
    }
  }

  UnifyDictionaryValues visitor{pool, value_type, dict_types, &values, out_transpose_maps };
  RETURN_NOT_OK(VisitTypeInline(value_type->id(), &visitor));

```
</details>

File: cpp/src/arrow/builder-dict.cc
Comment (fsaintjacques) [inline cpp/src/arrow/builder-dict.cc]: use expression with either `INT8_MAX` or `std::numeric_limits...`

File: cpp/src/arrow/builder-dict.cc
Comment (pitrou) [reply cpp/src/arrow/builder-dict.cc]: Thanks for the suggestion! I'll give it a try :-)

File: cpp/src/arrow/builder-dict.cc
Comment (pitrou) [reply cpp/src/arrow/builder-dict.cc]: I don't know, I find it as much or more readable like this.

Comment (bkietz) [reply cpp/src/arrow/builder-dict.cc]: I think this falls under "magic number", though: even if everybody knows what it is and finds it more readable, it's probably better to be explicit with a symbolic constant

Comment (pitrou) [reply cpp/src/arrow/builder-dict.cc]: Ok...

**Comment by pitrou** [comment]
This PR is ready for review now.

File: cpp/src/arrow/array-dict-test.cc
Comment (fsaintjacques) [inline cpp/src/arrow/array-dict-test.cc]: Can you type them `DictionaryArray` at the declaration to avoid all casts?

File: cpp/src/arrow/util/int-util-test.cc
Comment (fsaintjacques) [inline cpp/src/arrow/util/int-util-test.cc]: src.size() instead of 6, just in case.

File: cpp/src/arrow/util/int-util.cc
Comment (fsaintjacques) [inline cpp/src/arrow/util/int-util.cc]: Clang does this unrolling naturally, gcc does not :( https://godbolt.org/z/ZHobe1

File: cpp/src/arrow/util/int-util.cc
Comment (fsaintjacques) [inline cpp/src/arrow/util/int-util.cc]: Aren't you concerned by truncation due to incompatible size, maybe a static_assert somewhere?

**Review by fsaintjacques — APPROVED** [review]
LGTM

File: cpp/src/arrow/array-dict-test.cc
Comment (pitrou) [reply cpp/src/arrow/array-dict-test.cc]: Well, other APIs take `Array` pointers, not `DictionaryArray`.

File: cpp/src/arrow/util/int-util-test.cc
Comment (pitrou) [reply cpp/src/arrow/util/int-util-test.cc]: It may emit a cast warning, though :-/

File: cpp/src/arrow/util/int-util.cc
Comment (pitrou) [reply cpp/src/arrow/util/int-util.cc]: Neither does MSVC. The compiler generally doesn't have enough context to know whether unrolling is a good idea (I'm surprised than clang does it at `-O2`, btw). I'd rather unroll explicitly where _we_ have that context.

File: cpp/src/arrow/util/int-util.cc
Comment (pitrou) [reply cpp/src/arrow/util/int-util.cc]: What would we `static_assert`? Indeed it's up to the caller to ensure the input is ok (it is by construction in case of dictionary unification).

Comment (fsaintjacques) [reply cpp/src/arrow/util/int-util.cc]: sizeof(in_type) <= sizeof(out_type)
is_signed(in_type) == is_signed(out_type)

Comment (pitrou) [reply cpp/src/arrow/util/int-util.cc]: Ah, but the converse is supported as well.

**Review by bkietz — APPROVED** [review]

File: cpp/src/arrow/type.h
Comment (pitrou) [reply cpp/src/arrow/type.h]: Not in this implementation, but the XXX is about the more general case (all transpositions could actually be the identity function).

**Comment by wesm** [comment]
Giving this a look before merging, bear with me

File: cpp/src/arrow/array.cc
Comment (wesm) [inline cpp/src/arrow/array.cc]: Note: this is a specific case of the "take" function, a la ndarray.take or "from" verb in J. Here we have

out = in TAKE transpose_map

https://issues.apache.org/jira/browse/ARROW-772

File: cpp/src/arrow/array.cc
Comment (wesm) [inline cpp/src/arrow/array.cc]: Do you think it's worth the (minor) expense of checking for no-op case where `transpose_map` is `range(dictionary_size)`? This covers the case where the unified dictionary type is an extended version of the current type, e.g. as the result of a delta dictionary

Perhaps let's create a follow up JIRA so as to not hold up this patch

File: cpp/src/arrow/array.h
Comment (wesm) [inline cpp/src/arrow/array.h]: I think we do want that but how we handle it may require some care. This also came up in the context of Arrow Flight where the dictionaries are unknown at the start of an RPC session

https://issues.apache.org/jira/browse/ARROW-3144

I would suggest that we address both cases through a single design, i.e. a `MutableDictionaryType` subtype or something similar. 

One complexity that this introduces (which we need to be prepared for anyway) is the some code will need to check whether `type->dictionary()` is null or not (if it were null, it would mean that some data has been used inappropriately before the dictionary has been resolved), but I think that's OK.

File: cpp/src/arrow/array/builder_dict.cc
Comment (wesm) [inline cpp/src/arrow/array/builder_dict.cc]: This is a bit tricky. We should consider it in a follow up JIRA.

You could conceivably have a sequence of dictionaries like

```
a < b < c
a < b < c < d
a < b < c < d < e < f
```

This could result from a sequence of ordered delta dictionaries. You could argue in this case that the output type should also be ordered.

File: cpp/src/arrow/array/builder_dict.cc
Comment (wesm) [inline cpp/src/arrow/array/builder_dict.cc]: We should also consider the case where we want to conform a dictionary to an other, without unification. For example:

```
[a, b, c, a, b, c], dictionary [a, b, c]

conform to dictionary [a, c, d]

yields

[a, null, c, a, null, c]
```

This isn't unification so much as conformance. This was what was meant by my comment "Conform the dictionary indices to reference a common dictionary" in ARROW-554. The use case would be implementing casts between dictionary types -- this would include a safe and unsafe mode (safe would error if a value is made null)

Since this use case is not addressed in this patch (unless I missed it), let us create a follow up JIRA

File: cpp/src/arrow/type-test.cc
Comment (wesm) [inline cpp/src/arrow/type-test.cc]: I'm quite glad we have this function now =) We might consider adding even more testing sugar so that we can write `jsarray::int64("[3, 4, 5, 6]")` or some such

File: python/pyarrow/tests/test_types.py
Comment (wesm) [inline python/pyarrow/tests/test_types.py]: +1

**Review by wesm — COMMENTED** [review]
Thank you for kicking off this work. This can be a nuanced topic -- there's a number of follow up items that I've noted in comments. I'm going to go ahead and merge this patch and then we can create follow up JIRAs for discussion and work. A number of these items of follow up work will be forced by the development of other things, so no specific urgency

File: cpp/src/arrow/array.cc
Comment (pitrou) [reply cpp/src/arrow/array.cc]: There is already an XXX for that in `type.h`. I agree it could be beneficial in many cases (such as when you are simply augmenting an existing dictionary).

File: cpp/src/arrow/array.cc
Comment (pitrou) [reply cpp/src/arrow/array.cc]: Yes, though this one has transparent up/downcasting as well (which we probably don't want in the general case of a "take" function).

File: cpp/src/arrow/array/builder_dict.cc
Comment (pitrou) [reply cpp/src/arrow/array/builder_dict.cc]: IMHO the "ordered" flag doesn't seem to make sense. A dictionary is simply a linear lookup table, so it's always ordered in a sense.

File: cpp/src/arrow/array/builder_dict.cc
Comment (pitrou) [reply cpp/src/arrow/array/builder_dict.cc]: Hmm, I see. I have misinterpreted the original JIRA then. In any case, I was thinking of the CSV use case, where we'll be unifying dictionaries at the end.

File: cpp/src/arrow/array/builder_dict.cc
Comment (wesm) [reply cpp/src/arrow/array/builder_dict.cc]: ordered is part of the protocol https://github.com/apache/arrow/blob/master/format/Schema.fbs#L236

Comment (bkietz) [reply cpp/src/arrow/array/builder_dict.cc]: Minor optimization: if dictionaries have an ordering, that could be used to merge them without building a memo table.

File: cpp/src/arrow/type-test.cc
Comment (bkietz) [reply cpp/src/arrow/type-test.cc]: `"[3, 1, 4, 2]"_array(int64())`

File: cpp/src/arrow/array/builder_dict.cc
Comment (pitrou) [reply cpp/src/arrow/array/builder_dict.cc]: @bkietz I'm not sure I understand your suggestion. "ordered" here does not mean "sorted", just that the dictionary's order is semantically meaningful.

Comment (bkietz) [reply cpp/src/arrow/array/builder_dict.cc]: @pitrou My understanding is that if dictionaries to be unified are ordered, then the unified dictionary must preserve the order of shared emements. This in turn requires that those shared elements be identically ordered in each of the input dictionaries. For example, it seems impossible to unify `[0, 1], ordered` and `[1, 0], ordered` since the first claims it is semantically significant that 0 precedes 1 while the second claims the opposite. If shared elements are identically ordered then the new dictionary can be derived using longest common subsequence

Comment (bkietz) [reply cpp/src/arrow/array/builder_dict.cc]:
```
A   0     1 2 3 4 5       6

B     7 8 1   3 4   10 11 6

U   0 7 8 1 2 3 4 5 10 11 6
```

Comment (pitrou) [reply cpp/src/arrow/array/builder_dict.cc]: Right, but how does the algorithm look like and why would it more efficient than a series of hash lookups?

Comment (bkietz) [reply cpp/src/arrow/array/builder_dict.cc]: Sorry, you're right; talking about optimization is not justified when there is not a virgin which preserves order using a hash table. I'll write an example of each and benchmark them

Comment (wesm) [reply cpp/src/arrow/array/builder_dict.cc]: I would say it's not worth spending time right now on handling multiple "ordered" dictionaries (where the only case where "ordered" would be preserved is when all the dictionaries are prefixes of a common dictionary)

File: cpp/src/arrow/array.h
Comment (wesm) [reply cpp/src/arrow/array.h]: https://issues.apache.org/jira/browse/ARROW-4095

File: cpp/src/arrow/array.cc
Comment (wesm) [reply cpp/src/arrow/array.cc]: https://issues.apache.org/jira/browse/ARROW-4095

File: cpp/src/arrow/array.h
Comment (wesm) [reply cpp/src/arrow/array.h]: I left a comment in ARROW-3144 directing to this discussion

File: cpp/src/arrow/array/builder_dict.cc
Comment (wesm) [reply cpp/src/arrow/array/builder_dict.cc]: I created https://issues.apache.org/jira/browse/ARROW-4096. This is low priority

File: cpp/src/arrow/array/builder_dict.cc
Comment (wesm) [reply cpp/src/arrow/array/builder_dict.cc]: I opened https://issues.apache.org/jira/browse/ARROW-4097

**Comment by wesm** [comment]
OK I think I have all the follow up items accounted for in JIRA

**Pull request closed without merging**
